kernel: support for RTL8367C/S switch

namiltd opened the pull request:
PR https://github.com/openwrt/openwrt/pull/14599 compatible with main branch.

Support for RTL8367C/S Giga switch (RTL8367S, RTL8367RB-VB) without DSA.
Created to add support for routers like TP-Link Archer C5 v4 (PR https://github.com/openwrt/openwrt/pull/14418  and https://github.com/openwrt/openwrt/pull/14314)

luizluca commented on target/linux/generic/files/drivers/net/phy/rtl8367b.c:
```diff
@@ -234,6 +235,7 @@
 
 #define RTL8367B_CPU_PORT_NUM		5
 #define RTL8367B_NUM_PORTS		8
+#define RTL8367B_NUM_PHYS		5
```
The family of switches supported by rtl8367c includes 10 port switches (8+2ext). For the record, they are:
- RTL8363NB
- RTL8364NB
- RTL8365MB-VC
- RTL8367RB-VB
- RTL8367SB
- RTL8367S
- RTL8366SC
- RTL8363SC
- RTL8370MB
- RTL8310SR
- RTL8363NB-VB
- RTL8364NB-VB
- RTL8363SC-VB

Any switch that has more than 5 user ports will break this limit. The correct way to support multiple devices is to not use it as a constant.

For now, if it doesn't break current devices, we could stick with the constant but add a comment it might need to be revisited.

namiltd replied:
You are right, this patch focuses on 5+2ext .For larger/smaller switches the code needs to be reworked.

luizluca commented on target/linux/generic/files/drivers/net/phy/rtl8367b.c:
```diff
@@ -716,33 +734,46 @@ static int rtl8367b_write_phy_reg(struct rtl8366_smi *smi,
-		initvals = rtl8367r_vb_initvals_0;
-		count = ARRAY_SIZE(rtl8367r_vb_initvals_0);
-		break;
+	if (smi->flags & RTL8367B_FLAG_MATCH_CHIP_NUM) {
```
If we are introducing the chip_id match because of rtl8367c family, why not use that instead of using a flag? All devics using the compatible realtek,rtl8367c should use the chip_id match.

Edit: oh... flag came from the match data pointer. It is not pretty to use a pointer as a casted u32 bitmap (which will break with larger pointer size). It is way better to use a const struct with all required info in a readable form. That's the idea of the matching data.

Anyway, we are detecting the chip twice (detect and initregs). It should have save that info and just reuse it here.

namiltd replied:
I was hoping that it would be possible to base rtl8167d support on this code. In the case of the Tp-link router with Archer C5 v4 which comes in two versions rtl8167c and rtl8167d. If set compatible to realtek, rtl8367c will not work rtl8367d. For this reason, I wanted further detection to occur within the realtek,rtl8367c

luizluca commented on target/linux/generic/files/drivers/net/phy/rtl8367b.c:
```diff
@@ -1540,14 +1581,26 @@ static int rtl8367b_detect(struct rtl8366_smi *smi)
+		case 0x0276:
+		case 0x0597:
+		case 0x6367:
+			chip_name = "8367C";
```
This is not a chip name. rtl8367c is the vendor api version (as there are rtl8367b, rtl8367d). Please, use the corresponding chip name for each chip_num. You need to match chip_id with chip_ver to match the chip_name. You can check some known models at https://github.com/torvalds/linux/blob/master/drivers/net/dsa/realtek/rtl8365mb.c#L523

luizluca commented on target/linux/generic/files/drivers/net/phy/rtl8367b.c:
```diff
@@ -1580,15 +1633,36 @@ static struct rtl8366_smi_ops rtl8367b_smi_ops = {
+	{ .compatible = "realtek,rtl8367b" },
+	{
+		.compatible = "realtek,rtl8367c",
+		.data = (void *)RTL8367B_FLAG_MATCH_CHIP_NUM,
```
This is not how .data should be used. Point to a const struct and not abuse the pointer as an int.

luizluca reviewed (commented):
The rtl8367c support is not usable without the set phy addr for switch. I guess it should be squashed into the previous one or this one should come as the first patch. You could also use the matching data to set that info.

luizluca commented on target/linux/generic/files/drivers/net/phy/rtl8367b.c:
```diff
@@ -1634,21 +1634,36 @@ static struct rtl8366_smi_ops rtl8367b_smi_ops = {
+	.flags = RTL8367B_FLAG_MATCH_CHIP_NUM,
+};
+
+static const struct of_device_id rtl8367bc_match[] = {
```
rtl8367bc_match name makes no sense. Use rtl8367b_match as it is the module name.

luizluca commented:
These three commits do not stand by themselves. A commit is not the steps you took to reach a goal but the division of changes that even if you stop in the middle of a series, it should still work and give some benefits. You cannot introduce a broken code (like the pointer casted into a 32-bit int) and fix it in a following commit. For this PR, I would simply squash all commits into a single one.

I know realtek already do their job confusing api/chip/generation names. But we can do better, like in this PR title. I guess rtl8367c/s should be simply rtl8367s, which is actually what you are trying to do. rtl8367c is a much broader list of chips that will not work with this driver as it is.

luizluca commented:
You are force-pushing too frequently and the current diff is incomplete. Send a ping when you finished working on it.

namiltd commented:
> You are force-pushing too frequently and the current diff is incomplete. Send a ping when you finished working on it.

Current work completed

DragonBluep reviewed (commented):
Please reorganize the commit message. It looks horrible.

namiltd commented on target/linux/generic/files/drivers/net/phy/rtl8367b.c:
```diff
@@ -1540,14 +1576,29 @@ static int rtl8367b_detect(struct rtl8366_smi *smi)
-	case 0x1010:
-		chip_name = "8367R-VB";
+	switch (chip_num) {
+	case 0x6367:
```
The number of switch/case results from the fact that the code is the basis for expansion to additional chips.

namiltd commented on target/linux/generic/files/drivers/net/phy/rtl8367b.c:
```diff
@@ -716,30 +734,38 @@ static int rtl8367b_write_phy_reg(struct rtl8366_smi *smi,
-		initvals = rtl8367r_vb_initvals_0;
-		count = ARRAY_SIZE(rtl8367r_vb_initvals_0);
+	switch (chip_num) {
+	case 0x6367:
```
The number of switch/case results from the fact that the code is the basis for expansion to additional chips.

namiltd commented on target/linux/generic/files/drivers/net/phy/rtl8367b.c:
```diff
@@ -1021,6 +1047,16 @@ static int rtl8367b_setup(struct rtl8366_smi *smi)
 			RTL8367B_PORT_MISC_CFG_EGRESS_MODE_ORIGINAL <<
 				RTL8367B_PORT_MISC_CFG_EGRESS_MODE_SHIFT);
 
+	for (i = 0; i < RTL8367B_NUM_PHYS; i++) {
```
Removed in the next force-push, probably already done by the bootloader.

namiltd commented on target/linux/generic/files/drivers/net/phy/rtl8366_smi.c:
```diff
@@ -1547,6 +1547,9 @@ int rtl8366_smi_probe_of(struct platform_device *pdev, struct rtl8366_smi *smi)
 	}
 
+	if (of_property_read_u32(np, "phy-id", &smi->phy_id))
+		smi->phy_id = MDC_REALTEK_PHY_ADDR;
```
This name is original from rtl8366_smi.c

DragonBluep commented on target/linux/generic/files/drivers/net/phy/rtl8367b.c:
```diff
@@ -258,6 +258,8 @@
 	 RTL8367B_PORT_3 | RTL8367B_PORT_4 | RTL8367B_PORT_E1 |	\
 	 RTL8367B_PORT_E2)
 
+#define RTL8367B_FLAG_MATCH_CHIP_NUM	BIT(0)
```
This line can be removed.

namiltd replied:
Done

DragonBluep commented on target/linux/generic/files/drivers/net/phy/rtl8366_smi.h:
```diff
@@ -64,6 +64,7 @@ struct rtl8366_smi {
 	u8			dbg_vlan_4k_page;
 #endif
 	struct mii_bus		*ext_mbus;
+	int			phy_id;
```
`u32` or `unsigned int`. And I suggest moving it up.

namiltd replied:
Done

namiltd committed:
kernel: rtl8366_smi: explicitly set phy addr for switch

By default rtl8366_smi use phy addr 0 at mii-bus to access switch registers.
This patch allow to set it explicitly in dts-file:

	rtl8367 {
		compatible = "realtek,rtl8367b";
		phy-id = <29>; /* switch address at mii-bus */
		realtek,extif2 = <1 0 1 1 1 1 1 1 2>;
		mii-bus = <&mdio>;
		cpu-port = <7>;
	}

Use default 0 address if not set.
Backward compatibility tested on tplink archer c2 v1 (rtl8367rb switch)

Signed-off-by: Serge Vasilugin <vasilugin@yandex.ru>
[code style fixes, add explicit phy_id assignment in probe_plat, use
phy-id instead of phy_id for of property name]
Signed-off-by: Chuanhong Guo <gch981213@gmail.com>
[rebase]
Signed-off-by: Gaspare Bruno <gaspare@anlix.io>
[added phy_id to struct rtl8366_smi]
Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

namiltd committed:
kernel: support for RTL8367C/S switch

From driver point of view no differance between rtl8367b and rtl8367s
if it connected through EXT2 (rgmii only).
So this trivial patch add some identification and initialization only.
SGMII/HSGMII mode for EXT1 is not implemented for the sake of patch
clairity.

Signed-off-by: Serge Vasilugin <vasilugin@yandex.ru>
[Fix code format]
Signed-off-by: DENG Qingfang <dengqf6@mail2.sysu.edu.cn>
[add flags to separate chip_num/chip_id detection; drop error print in
rtl8367b_init_regs, drop unnecessary info prints, code style fixes]
Signed-off-by: Chuanhong Guo <gch981213@gmail.com>
[rebase; use MII macros]
Signed-off-by: Gaspare Bruno <gaspare@anlix.io>
[code optimization]
Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

luizluca commented:
FYI, the rtl8367c chips can handle a smaller clock delay in SMI read/write. You can use smi->clk_delay = 10; for those chips. However, that is not a major feature as reading/writing registers is not too frequent.

Leo-PL commented:
@namiltd how did you determine the phy-id for RTL8637S? I'm trying to get RTL8363S working on Ruckus ZF7363 and it seems that phy-id 0 is wrong for it, too. And this switch seems to be of the same family, just less ports.
That's what I get on failed probe:
```
[    5.075411] rtl8367b 1a000000.eth:switch: using MDIO bus 'ag71xx_mdio'
[    5.082577] rtl8367b 1a000000.eth:switch: unknown chip num:0020 ver:0020, mode:0020
[    5.090226] rtl8367b 1a000000.eth:switch: chip detection failed, err=-19
```
mdio-tools shows this to me:
```
root@ruckus-zf7363:/sys/bus/mdio_bus/devices# mdio 
fixed-0
mdio.0
root@ruckus-zf7363:/sys/bus/mdio_bus/devices# mdio mdio.0
 DEV      PHY-ID  LINK
0x00  0x001cc930  down
0x01  0x001cc930  down
0x02  0x001cc981  up
0x1e  0x01410e40  up
root@ruckus-zf7363:/sys/bus/mdio_bus/devices# mdio mdio.0 2
BMCR(0x00): 0x2100
  flags: -reset -loopback -aneg-enable -power-down -isolate -aneg-restart
         -collision-test
  speed: 100-full

BMSR(0x01): 0x7805
  capabilities: -100-t4 +100-tx-f +100-tx-h +10-t-f +10-t-h -100-t2-f -100-t2-h
  flags:        -ext-status -aneg-complete -remote-fault -aneg-capable +link
                -jabber +ext-register

ID(0x02/0x03): 0x001cc981
```
Please disregard 0x1e, this is an unrelated Marvell PHY handling eth0.

namiltd commented:
mt7620a_tplink_archer-c5-v4.dts:
```
	rtl8367s {
		compatible = "realtek,rtl8367b";
		cpu_port = <7>;
		realtek,extif2 = <1 0 1 1 1 1 1 1 2>;
		mii-bus = <&mdio0>;
		phy-id = <29>;
	};

namiltd commented:
For information: this patch can be backported to version 23.05 without any changes.

Leo-PL commented:
Okay, but from where did that `phy-id = <29>` come from?
I have no GPL sources from Ruckus to dig that out, unfortunately.

DragonBluep commented:
> Okay, but from where did that `phy-id = <29>` come from?

@Leo-PL I think it can only be 0 or 29. https://github.com/search?q=%22PHY+ID+0+or+29%22&type=code

> [    5.082577] rtl8367b 1a000000.eth:switch: unknown chip num:0020 ver:0020, mode:0020
> [    5.090226] rtl8367b 1a000000.eth:switch: chip detection failed, err=-19

This driver doesn't support your switch chip num `0x0020`. Perhaps you can add this id based on this PR.

Leo-PL commented:
Yeah, I guess i'll try 29 blindly and see what happens. It for sure doesn't - no open-source driver for RTL8363S is out there yet, be it swconfig or DSA, but I'll try to take care of that based on what I've read in Realtek SDK for it.

Leo-PL commented:
```
[    5.075406] rtl8367b 1a000000.eth:switch: using MDIO bus 'ag71xx_mdio'
[    5.082569] rtl8367b 1a000000.eth:switch: unknown chip num:ffff ver:ffff, mode:ffff
[    5.090218] rtl8367b 1a000000.eth:switch: chip detection failed, err=-19
```
so nothing again - I guess I need to reload stock firmware and log the MDIO traffic somehow - glad that I can probably find these pins on Marvell PHY.

namiltd commented:
@Leo-PL Maybe try done loop for phy_id from 0 to 255

luizluca commented:
Finally we got rtl8367c supported! Thanks to all involved!
